docs/1.0. Authorization Practice.md

```diff
@@ -71,6 +82,23 @@ For example, access to the following subscription would require token claims as
 }
 ```
 
+### IS-05 - Device Connection Management
+
+#### Transport File Authorization
+
+Where an IS-05 Sender exposes a transport file via the `/single/senders/{sender_id}/transportfile` path, this MUST NOT
+use HTTP redirects when the API is operating in authorized mode. The API MUST return an HTTP 200 code as opposed to a
+3xx code when the request is successful.
```

Comment:
I am not convinced this is a hard MUST for things to work 🤔 you would need to be authorized regardless of the status code, otherwise a 4xx would be returned

Reply:
Are you suggesting that redirects should be permitted or that even when not redirecting a different code to 200 could be returned on success?

Reply:
I apologize, short comment following up from the linked issue, https://github.com/AMWA-TV/nmos-authorization-practice/issues/5#issuecomment-715577119

I believe we can allow redirects in authorized mode, it should not introduce any vulnerability to my knowledge.

Reply:
The concern was less about vulnerability and perhaps more about knowing what your token needs to contain ahead of making a request (and avoiding inconsistency in claims required by different implementations as a result).

If I make a request to IS-05's `/transportfile` and receive a redirect to either another path or something under the IS-04 namespace for example, I then need to make sure I have the relevant permissions to access that resource too. It seems possible that a controlling device would then need to make another round trip to the Authorization Server in order to request a suitable token before following the redirect.

The redirect could of course forward to a path that doesn't require authorization, but my feeling was that we should be writing things to try and ensure that an end user can't unknowingly end up with some paths which remain insecure as they're masked by a redirect.

Reply:
I think the concern was that redirecting to URLs outside of **/x-nmos** could introduce more interoperability issues once auth is considered. But @prince-chrismc showed me it's perhaps not as much of a problem as I thought.

The flow would be:

NMOS-aware client sends request to **/x-nmos/connection/.../transportfile** with token with `connection` scope and appropriate `x-nmos-connection` private claim.

The API checks the credentials and sends a 3xx response with `Location` response header with a URL, e.g. **/x-manufacturer/.../video.sdp**.

At this point, the client has to make a request to that manufacturer-specific URL without auth, unless it knows better by some out-of-band means (not good for interop). The token that worked for the IS-05 API is not likely to work for the manufacturer-specific URL and in fact many HTTP client frameworks that support automatic redirect will always strip auth request headers.

Assuming the manufacturer-specific URL does require auth, it will respond with 401 Unauthorized and a `WWW-Authenticate` header to decide how to retry. That response header can (should...) indicate the required `scope` and some human-readable information, but it doesn't programmatically go into the detail of `claims` etc.

So if as an API implementation you return 3xx, you're going to have to deal with at least three requests to provide the SDP file, rather than just 1 if you return it directly with a 200 OK.

Chris pointed out that a similar issue is the case for trailing slashes and redirects, as I also noted in  https://github.com/AMWA-TV/nmos-authorization-practice/issues/5#issuecomment-709015268.

The difference there is that the client could examine the redirect `Location` and see that it's still a **/x-nmos/{same api}/{same version}...** URL and send the same token - provided it takes ownership of doing redirects itself rather than letting a framework do it itself. Then it's still two requests, but no worse than it was before we added HTTPS and auth?.

Bottom line is would it be reasonable to change it to "**SHOULD** return 200 OK, as opposed to 3xx code when the request is successful."

Reply:
Apologies for the long comment 😄

Reply:
Thank you @garethsb-sony for the summary! 

I understand much better now where the interoperability issue would come into play. My original train of thought assumed auth was not required or _easier_ to resolve, but there's still an ambiguity.

Thinking more generally about redirects, perhaps there the BCP should offer stricter guidance?

Reply:
Reviewed on call. The MUST in line 90 will stay. PR will be merged and a new issue created for redirect guidance – that may mean a SHOULD in IS-10 and a MUST in BCP-003-02.